lib/common/redis.c

```diff
@@ -38,21 +59,60 @@ static void on_redis_connect(const redisAsyncContext *redis, int status)
         conn->state = H2O_REDIS_CONNECTION_STATE_CLOSED;
         conn->_redis = NULL;
         if (conn->on_close != NULL) {
-            conn->on_close(redis->errstr);
+            conn->on_close(redis->c.errstr);
         }
     }
+
+    if (h2o_timeout_is_linked(&conn->_connect_timeout_entry)) {
+        h2o_timeout_unlink(&conn->_connect_timeout_entry);
+        h2o_timeout_dispose(conn->loop, &conn->_connect_timeout);
+    }
 }
 
-static void on_redis_disconnect(const redisAsyncContext *redis, int status)
+static void close_connection(h2o_redis_conn_t *conn)
 {
-    h2o_redis_conn_t *conn = (h2o_redis_conn_t *)redis->data;
     conn->state = H2O_REDIS_CONNECTION_STATE_CLOSED;
-    conn->_redis = NULL;
+
     if (conn->on_close != NULL) {
-        conn->on_close(redis->errstr);
+        conn->on_close(conn->_redis->c.errstr);
+    }
+    conn->_redis = NULL;
+
+    if (h2o_timeout_is_linked(&conn->_connect_timeout_entry)) {
+        h2o_timeout_unlink(&conn->_connect_timeout_entry);
+    }
+    if (h2o_linklist_is_linked(&conn->_connect_timeout._link)) {
+        h2o_timeout_dispose(conn->loop, &conn->_connect_timeout);
     }
 }
 
+static void on_disconnect(const redisAsyncContext *redis, int status)
+{
+    h2o_redis_conn_t *conn = (h2o_redis_conn_t *)redis->data;
+    close_connection(conn);
+}
+
+static void on_connect_timeout_deferred(h2o_timeout_entry_t *entry)
+{
+    h2o_redis_conn_t *conn = H2O_STRUCT_FROM_MEMBER(h2o_redis_conn_t, _defer_timeout_entry, entry);
+
+    assert((conn->_redis->c.flags & REDIS_CONNECTED) == 0);
```

Comment:
Am I correct in assuming that there is a value that represents the connection state in the `redisAsyncContext`? If so, can we use the value instead of having `h2o_redis_connection_state_t`?

Having two states that represents the same thing is a source of confusion.

Reply:
> there is a value that represents the connection state in the redisAsyncContext?

Yes, but it's a bit redundant and seems to be not intended to used directly from outside of hiredis. And also considering that we have a bit different state model described in  https://github.com/h2o/h2o/pull/1152#discussion_r164332530, I think we should have our own state representation.